rawfile/utils/task_manager.py

```diff
@@ -0,0 +1,195 @@
+                TaskName(failed_task["task"]),
+                *failed_task["args"],
+                **failed_task["kwargs"],
+            )
+
+    def hash_task_info(self, task: TaskName, *args, **kwargs):
+        raw = json.dumps(kwargs) + json.dumps(args) + task.value
+        return hashlib.md5(raw.encode()).hexdigest()
+
+    def get_tasks(
+        self, state: TaskState | None = None, retriable: bool | None = None
+    ) -> dict[str, TaskInfo]:
+        with self._lock:
+            with open(f"{consts.DATA_DIR}/tasks.json") as tasks_file:
+                data = json.loads(tasks_file.read() or "{}")
+                if retriable is not None:
+                    if retriable:
+                        data = {
+                            k: v
+                            for k, v in data.items()
+                            if v["retry_count"] < self._max_retry
+                            and v["state"] == TaskState.FAILED
+                        }
+                    else:
+                        data = {
+                            k: v
+                            for k, v in data.items()
+                            if v["retry_count"] >= self._max_retry
+                        }
+                if state is not None:
+                    data = {k: v for k, v in data.items() if v["state"] == state}
+                return data
+
+    def get_task(self, task_id) -> None:
+        self.get_tasks().get(task_id, None)
+
+    def remove_task(self, task_id: str):
+        logger.debug("Removing task", task_id=task_id)
+        with self._lock:
+            with open(self._tasks_store_path, "r+") as tasks_file:
+                current: dict[str, TaskInfo] = json.loads(tasks_file.read() or "{}")
+                current.pop(task_id)
+                tasks_file.truncate(0)
+                tasks_file.seek(0)
+                json.dump(current, tasks_file)
+
+    def done_callback(self, task_id: str):
+        state: TaskState
+        task_info: TaskInfo = self._tasks.get(task_id, {}).get("info", {})
+        future: Future | None = self._tasks.get(task_id, {}).get("future", None)
+        if not (future and (future.exception() or future.cancelled())):
+            state = TaskState.COMPLETED
+            logger.success("Task Completed", task_id=task_id, task=task_info)
+        else:
+            exc = (
+                future.exception()
+                if future.exception()
+                else Exception("Task was cancelled")
+            )
+            state = TaskState.FAILED
+            logger.opt(exception=exc).error(
+                "Task failed", task_id=task_id, task=task_info, exception=exc
+            )
+        task_info["state"] = state
+        self.save_task(task_id, task_info)
+
+    def save_task(self, task_id: str, task_info: TaskInfo):
+        with self._lock:
+            logger.debug("Saving task", task_id=task_id, task=task_info)
+            with open(self._tasks_store_path, "r+") as tasks_file:
```

Comment:
would it be safer to have unique files per task rather than a single for all?

Reply:
Handling a specific file per task will be complicated, I know we don't need locking of task file in some places anymore, but it will be complicated I think

Reply:
Yeah possibly, was just wondering about write atomicity, and what happens if the node crashes half-way through the write
Maybe just the fsync for now is fine